compiler/backend/src/org/jetbrains/kotlin/codegen/JvmCodegenUtil.java

```diff
@@ -199,15 +199,27 @@ public static boolean couldUseDirectAccessToProperty(
         if (KotlinTypeMapper.isAccessor(property)) return false;
 
         CodegenContext context = contextBeforeInline.getFirstCrossInlineOrNonInlineContext();
-        // Inline functions can't use direct access because a field may not be visible at the call site
-        if (context.isInlineMethodContext()) {
+        // Inline functions or inline classes can't use direct access because a field may not be visible at the call site
+        if (context.isInlineMethodContext() || (context.getEnclosingClass() != null && context.getEnclosingClass().isInline())) {
             return false;
         }
 
+        DeclarationDescriptor propertyContainingDeclaration = property.getContainingDeclaration();
+        boolean isCompanionPropBackedInOuter = DescriptorUtils.isCompanionObject(propertyContainingDeclaration) &&
+                                               JvmAbi.isPropertyWithBackingFieldInOuterClass(property);
         if (!isCallInsideSameClassAsDeclared(property, context)) {
             if (!isDebuggerContext(context)) {
-                // Unless we are evaluating expression in debugger context, only properties of the same class can be directly accessed
-                return false;
+                // Property with backing field into an outer class could be used directly when access is done into the context of the outer
+                // class if the conditions further afield are respected, otherwise accessors must be used.
+                if (isCompanionPropBackedInOuter) {
+                    if (propertyContainingDeclaration.getContainingDeclaration() !=
```

Comment:
1. Because `propertyContainingDeclaration` is `ClassDescriptor` and `context.getContextDescriptor()` is a `FunctionDescriptor` and what I want to do is to check if the access is done into the same class than the owner of the property. Moreover `propertyContainingDeclaration` is the class containing the declaration but not the outer class containing the backed field.
2. `isCallInsideSameClassAsDeclared` does not take into account property with backed field into the outer class, the check is done with where the property is declared and not where the property is generated, which is not the same place for a property that has a backed field into the outer class.

Thanks to your comments, I think that the cleaner way is to remove code that I added and to update `isCallInsideSameClassAsDeclared` to take into account property with backed field.